unity/Assets/Scripts/ActionDispatcher.cs

```diff
@@ -0,0 +1,233 @@
+
+                    foreach(var p in method.GetParameters()) {
+                        if (actionParams.Contains(p.Name)) {
+                            matchCount++;
+                        }
+                    }
+                }
+
+                // preference is given to the method that matches all parameters for a method
+                // even if another method has the same matchCount (but has more parameters)
+                if (matchCount > bestMatchCount) {
+                    bestMatchCount = matchCount;
+                    matchedMethod = method;
+                }
+            }
+
+        }
+
+        return matchedMethod;
+    }
+
+    public static void Dispatch(System.Object target, dynamic serverCommand) {
+
+        MethodInfo method = getDispatchMethod(target, serverCommand);
+
+        if (method == null) {
+            throw new InvalidActionException();
+        }
+
+        List<string> missingArguments = null; 
+        System.Reflection.ParameterInfo[] methodParams = method.GetParameters();
+        object[] arguments = new object[methodParams.Length];
+        if (methodParams.Length == 1 && methodParams[0].ParameterType == typeof(ServerAction)) {
+            arguments[0] = serverCommand.ToObject(methodParams[0].ParameterType);
+        }  else {
+            for(int i = 0; i < methodParams.Length; i++) {
+                System.Reflection.ParameterInfo pi = methodParams[i];
+                // allows for passing in a ServerAtion as a dynamic to ProcessControlCommand
+                if (serverCommand.GetType() == pi.ParameterType){
+                    arguments[i] = serverCommand;
+                } else if (serverCommand.ContainsKey(pi.Name)) {
+                    arguments[i] = serverCommand[pi.Name].ToObject(pi.ParameterType);
+                } else {
+                    if (!pi.HasDefaultValue)  {
+                        if (missingArguments == null) {
+                            missingArguments = new List<string>();
+                        }
+                        missingArguments.Add(pi.Name);
+                    }
+                    arguments[i] = Type.Missing;
+                }
+            }
+        }
+        if (missingArguments != null) {
+            throw new MissingArgumentsActionException(missingArguments);
+        }
+
+        method.Invoke(target, arguments);
+    }
+
+
+
+}
+
+public class MethodParamComparer: IComparer<MethodInfo> {
+
+
+    public int Compare(MethodInfo a, MethodInfo b) {
+        int requiredParamCountA = requiredParamCount(a);
+        int requiredParamCountB = requiredParamCount(b);
+        int result = requiredParamCountA.CompareTo(requiredParamCountB);
+        if (result == 0) {
+            result = paramCount(a).CompareTo(paramCount(b));
+        }
+        return result;
+    }
+
+    private int paramCount(MethodInfo method) {
+        return method.GetParameters().Length;
+    }
+
+    private int requiredParamCount(MethodInfo method) {
```

Comment:
Make static

Reply:
Can you elaborate on this? I'm not sure how I would make this method static since its part of the comparer.

Reply:
Because it does not use any internal state of the `MethodParamComparer` instance, it is like a library function that just takes a `method` variable and returns a value, independent of the instance. It's a general practice to prefer static methods due to object oriented opacity, and makes it clear that the method is like a utility and not really doing anything with the internal state of the instance.

Reply:
A minor point really, just easy to make static